File: go/solver/solver.go
package solver

import (
	"sort"
)

// Board represents the sudoku board. uint8 allows a 2x faster processing (and we always have 0 <= 9 <= 255)
type Board = [9][9]uint8

func digitsPossible(S *Board, i uint8, j uint8) uint16 {
	if S[i][j] != 0 { // cell already set
		return 0
	}

	var digits uint16
	for n := uint8(1); n <= 9; n++ {
		if isAvailable(S, i, j, n) {
			digits |= 1 << n
		}
	}

	return digits
}

func matrixPossibilities(S *Board) [9][9]uint16 {
	tab := [9][9]uint16{}

	for i := uint8(0); i < 9; i++ {
		for j := uint8(0); j < 9; j++ {
			tab[i][j] = digitsPossible(S, i, j)
		}
	}
	return tab
}

type pos struct {
	i uint8
	j uint8
}

// tableauOrder returns the order in which the cells must be processed
func tableauOrder(S *Board) []pos {
	type posWithScores struct {
		pos   pos
		score int // as minimal as possible
	}
	tab := matrixPossibilities(S)

	// compute scores
	listeScores := make([]posWithScores, 0, 81)
	for i := uint8(0); i < 9; i++ {
		for j := uint8(0); j < 9; j++ {
			if tab[i][j] != 0 {
				listeScores = append(listeScores, posWithScores{
					pos:   pos{i: i, j: j},
					score: 100*countOnes(tab[i][j]) - alignedNeighbors(S, i, j) - squareNeighbors(S, i, j),
				})
			}
		}
	}

	// sort by score
	sort.SliceStable(listeScores, func(a, b int) bool {
		return listeScores[a].score < listeScores[b].score
	})
	liste := make([]pos, 0, len(listeScores))
	for _, p := range listeScores {
		liste = append(liste, p.pos)
	}
	return liste
}

////////////////////////////
//////   Résolution   //////
////////////////////////////

// not really block : only checks the 4 squares that can't be reached with col/row checking
func isAvailableInBloc(S *Board, i uint8, j uint8, n uint8) bool {
	for k := floor3(i); k < floor3(i)+3; k++ {
		if k == i {
			continue
		}
		for l := floor3(j); l < floor3(j)+3; l++ {
			if l != j && S[k][l] == n {
				return false
			}
		}
	}
	return true
}

func isAvailableInLine(S *Board, i uint8, j uint8, n uint8) bool {
	for k := uint8(0); k < 9; k++ {
		if (S[i][k] == n && k != j) || (S[k][j] == n && k != i) {
			return false
		}
	}
	return true
}

func isAvailable(S *Board, i uint8, j uint8, n uint8) bool {
	return isAvailableInLine(S, i, j, n) && isAvailableInBloc(S, i, j, n)
}

func floor3(n uint8) uint8 {
	return 3 * (n / 3)
}

type stats struct {
	Tries     int
	GoingBack int
}

// Solve solves a sudoku and returns the answer
func Solve(Si Board) (Board, stats) {
	S := &Si
	// Initialise possibilities, order and digit position
	possibilities := matrixPossibilities(S)
	sliceOrder := tableauOrder(S)
	maxDigitToFind := uint8(len(sliceOrder))
	var currentCandidate [9][9]uint8
	for i := uint8(0); i < 9; i++ {
		for j := uint8(0); j < 9; j++ {
			currentCandidate[i][j] = nextCandidate(0, possibilities[i][j])
		}
	}

	// variables
	var rank uint8
	var stats stats

	for rank < maxDigitToFind {
		stats.Tries++

		// Which cell must be processed ?
		n := sliceOrder[rank]
		i := n.i
		j := n.j

		if currentCandidate[i][j] != 0 {
			// There is a digit in the list of possibilities to put here

			if isAvailable(S, i, j, uint8(currentCandidate[i][j])) {
				// digit available -> go forward to next cell
				S[i][j] = uint8(currentCandidate[i][j])
				rank++
			} else {
				// digit already in line, col or 3x3 cell -> try higher digit for same cell
				currentCandidate[i][j] = nextCandidate(currentCandidate[i][j], possibilities[i][j])
			}
		} else {
			// there is no digit possible for this cell

			// first 'reset' the cell state
			S[i][j] = 0
			currentCandidate[i][j] = nextCandidate(0, possibilities[i][j])

			// then go back to previous cell
			rank--
			stats.GoingBack++
			n = sliceOrder[rank]
			i = n.i
			j = n.j
			// previous digit was available but we found it wasn't okay, so increase it
			currentCandidate[i][j] = nextCandidate(currentCandidate[i][j], possibilities[i][j])
		}

	}

	return *S, stats
}

File: rust/src/solver.rs
// Only used at init time, perf not important
fn digits_possible(s: [[u8; 9]; 9], i: usize, j: usize) -> Option<Vec<u8>> {
    if s[i][j] != 0 {
        // cell already set
        return None;
    }

    let mut digits = Vec::new();
    for n in 1..10 {
        if is_available(&s, i, j, n) {
            digits.push(n);
        }
    }

    Some(digits)
}

// Only used at init time, perf not important
fn matrix_possibilities(s: [[u8; 9]; 9]) -> [[Vec<u8>; 9]; 9] {
    let mut tab: [[Vec<u8>; 9]; 9] = Default::default();

    for (i, line) in tab.iter_mut().enumerate() {
        for (j, cell) in line.iter_mut().enumerate() {
            match digits_possible(s, i, j) {
                Some(v) => {
                    if !v.is_empty() {
                        *cell = v;
                    }
                }
                None => continue,
            }
        }
    }
    tab
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
struct Pos {
    i: usize,
    j: usize,
}

// Only used at init time, perf not important
fn tableau_order(s: [[u8; 9]; 9]) -> Vec<Pos> {
    #[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
    struct PosWithScores {
        pos: Pos,
        score: i32,
    }

    let possibilities = matrix_possibilities(s);

    let mut liste_scores = vec![];
    for (i, line) in possibilities.iter().enumerate() {
        for (j, cell) in line.iter().enumerate() {
            let score =
                100 * cell.len() as i32 - aligned_neighbors(&s, i, j) - square_neighbors(&s, i, j);
            if score <= 0 {
                continue;
            }
            liste_scores.push(PosWithScores {
                pos: Pos { i, j },
                score,
            });
        }
    }

    liste_scores.sort_by(|a, b| a.score.cmp(&b.score));
    liste_scores.into_iter().map(|p| p.pos).collect()
}

// Used at each step, perf is important
fn is_available_in_line(s: &[[u8; 9]; 9], i: usize, j: usize, n: u8) -> bool {
    for k in 0..9 {
        if (s[i][k] == n && k != j) || (s[k][j] == n && k != i) {
            return false;
        }
    }
    true
}

// Used at each step, perf is important
fn is_available_in_bloc(s: &[[u8; 9]; 9], i: usize, j: usize, n: u8) -> bool {
    for (k, line) in s.iter().enumerate().skip((i / 3) * 3).take(3) {
        if k == i {
            continue;
        }
        for (l, case) in line.iter().enumerate().skip((j / 3) * 3).take(3) {
            if l != j && *case == n {
                return false;
            }
        }
    }

    true
}

// Used at each step, perf is important
fn is_available(s: &[[u8; 9]; 9], i: usize, j: usize, n: u8) -> bool {
    is_available_in_line(s, i, j, n) && is_available_in_bloc(s, i, j, n)
}

pub struct Stats {
    tries: i32,
    going_back: i32,
}

// main function
pub fn solve(si: [[u8; 9]; 9]) -> ([[u8; 9]; 9], Stats) {
    let mut s = si;
    let possibilities = matrix_possibilities(s);
    let slice_order = tableau_order(s);
    let max_digit_to_find = slice_order.len();
    let mut index_of_current_digit_for: [[usize; 9]; 9] = [[0; 9]; 9];

    let mut rank: usize = 0;
    let mut stats = Stats {
        tries: 0,
        going_back: 0,
    };

    while rank < max_digit_to_find {
        stats.tries += 1;

        let n = slice_order[rank];
        let i = n.i;
        let j = n.j;

        if index_of_current_digit_for[i][j] < possibilities[i][j].len() {
            let client = possibilities[i][j][index_of_current_digit_for[i][j]];

            if is_available(&s, i, j, client) {
                s[i][j] = client;
                rank += 1;
            } else {
                index_of_current_digit_for[i][j] += 1;
            }
        } else {
            s[i][j] = 0;
            index_of_current_digit_for[i][j] = 0;
            rank -= 1;
            stats.going_back += 1;
            let n = slice_order[rank];
            let i = n.i;
            let j = n.j;
            index_of_current_digit_for[i][j] += 1;
        }
    }

    (s, stats)
}

fn aligned_neighbors(s: &[[u8; 9]; 9], i: usize, j: usize) -> i32 {
    let mut count = 0;
    for k in 0..9 {
        if j != k && s[i][k] != 0 {
            count += 1;
        }
        if i != k && s[k][j] != 0 {
            count += 1;
        }
    }

    count
}

fn square_neighbors(s: &[[u8; 9]; 9], i: usize, j: usize) -> i32 {
    let mut count = 0;
    let i_start = (i / 3) * 3;
    let j_start = (j / 3) * 3;
    for (k, _) in s.iter().enumerate().skip(i_start).take(3) {
        for l in j_start..j_start + 3 {
            if s[k][l] != 0 && (k != i || l != j) {
                count += 1;
            }
        }
    }

    count
}
